File: server/hardware_order_test.go
package main

import "testing"

func TestHardwareUsesTickStartPosition(t *testing.T) {
	for _, warp := range []bool{true, false} {
		s := initial()
		for i := range s.Players {
			s.Players[i].Stun = 100
		}
		s.Ball = Ball{X: 48 * (22.4 / 576), Z: -11.1, H: 1, VZ: -8, Owner: -1, LastTouch: 7, SpeedTimer: 100}
		if warp {
			s.Ball.X = 8
			s.Ball.Z = 11.1
			s.Ball.VZ = 8
		}
		s.step(simulationStep, [2]Input{})
		if s.Score[0] != 0 || warp && s.Ball.Z <= 11.2 || !warp && s.Ball.Z >= -11.2 {
			t.Fatal("premature contact")
		}
		s.step(simulationStep, [2]Input{})
		if warp {
			if s.Ball.Z >= 0 || s.Event.Kind != 12 {
				t.Fatal("missing warp")
			}
		} else if s.Score[0] != 2 {
			t.Fatal("missing star")
		}
	}
}

File: src/hardware-order.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { initial, step, simulationStep } from "./game.js";
test("warp and star contacts use the position at tick start", () => {
  for (const warp of [true, false]) {
    const s = initial();
    s.players.forEach((p) => (p.stun = 100));
    Object.assign(s.ball, {
      x: warp ? 8 : (48 * 22.4) / 576,
      z: warp ? 11.1 : -11.1,
      h: 1,
      vz: warp ? 8 : -8,
      owner: -1,
      lastTouch: 7,
      speedTimer: 100,
    });
    step(s, simulationStep, {}, [true, true]);
    assert.equal(s.score[0], 0);
    assert.ok(warp ? s.ball.z > 11.2 : s.ball.z < -11.2);
    step(s, simulationStep, {}, [true, true]);
    if (warp) {
      assert.ok(s.ball.z < 0);
      assert.equal(s.event.kind, 12);
    } else {
      assert.equal(s.score[0], 2);
    }
  }
});
